### PR TITLE
Remove Performance/CaseWhenSplat cop

### PR DESCRIPTION
**Pull request by urbanautomaton**
### TL;DR

The Performance/CaseWhenSplat cop provides marginal and context-sensitive performance benefits, and advises unsafe changes. We should remove it.

### Description

The Performance/CaseWhenSplat cop was [originally introduced](https://github.com/bbatsov/rubocop/pull/2143) to optimise memory allocations in case statements with splatted conditions. In profiling over some of the most case-statement-heavy parts of the rubocop codebase it showed a 2% improvement in allocations.

However, as the original PR notes, the suggested change:

* is only safe to make if the case conditions are disjoint
* doesn't confer performance benefits in all conditions

Contrary to this, the cop's docs state:

> Since Ruby does not support fall through inside of `case` `when`, like some other languages do, the order of the `when` branches does not matter.

This isn't true, and is likely to result in inexperienced users introducing bugs when complying with the cop. In the cop's examples, the following replacement is advised:

```ruby
# bad
case foo
when *condition
  bar
when baz
  foobar
end

# good
case foo
when baz
  foobar
when *condition
  bar
end
```

Here, if `baz` is a member of `condition`, this edit will alter the meaning of the case statement. In the original PR this wasn't considered a significant concern, but I'd argue that it's highly problematic for a linter to propose edits that have semantic implications.

In response we could:

* Update the cop's documentation
* Disable it by default
* Remove it

In this PR I've prepped the removal option, since I believe the performance gains of the cop are so small and context-sensitive that I don't believe it makes sense to implement it as a linter rule, particularly since we can't guarantee that the changes advised are safe.

I'd welcome feedback on what approach to take, however. What do you think?

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists). *N/A*
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Run `rake default` or `rake parallel`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by rrosenblum** [comment]
You are absolutely right about the potential change to functionality if the same condition is defined multiple times. In the time since the cop was released, 3 years ago, this is the first time anyone has brought this up as an issue. To me, this means the original assumption about it being low risk is correct. However, low risk does not mean no risk.

> In profiling over some of the most case-statement-heavy parts of the rubocop code base it showed a 2% improvement in allocations

This is a little misleading. This cop provided a 2% memory reduction for all of RuboCop. The memory consumption of `variable_force.rb` dropped by 30%. `variable_force.rb` was the main file that was impacted by this cop. The change to `variable_force.rb` was significant.

> Since Ruby does not support fall through inside of case when, like some other languages do, the order of the when branches does not matter.

What I meant here is that in some languages, if the same condition shows up in multiple `when` statements, all of the matching `when` conditions will be executed in the order they are hit. This is not the case with Ruby. Again an assumption is being made here that because of this, a user will want to make sure that each of there `when` conditions is unique. As you pointed out, this may not be true if someone is manually defining a `when` condition with a condition that also exists in a splat expanded variable.

Like many of the other performance related cops, the performance benefits are highly dependent on the context of the code itself. I feel like it is a bit extreme to completely remove this cop. I think a better course of action would be to update the documentation and/or disable it by default.

**Comment by urbanautomaton** [comment]
Hi @rrosenblum - thanks for the response. 👍 

I guess my feeling is that a linter should never recommend semantically significant changes to code, even if the likelihood of it causing problems is low - rubocop is ubiquitous in the ruby world, so if something can happen, it almost certainly will (and you may not hear about it, either).

For what it's worth, the code it objected to for me was something like the following:

```ruby
IGNORE_EXCEPTIONS = [OopsyDaisyError, NeverMindItllProbablyGetBetterOnItsOwnError]
SPECIAL_EXCEPTIONS = [PokedMyselfInTheEyeError, WhyDidIDoThatError]

def handle_error(error)
  case error
  when *IGNORE_EXCEPTIONS then return
  when *SPECIAL_EXCEPTIONS then do_a_special_thing(error)
  when StandardError then do_a_standard_thing(error)
  else do_a_catchall_thing(error)
  end
end
```

This doesn't feel particularly exotic, as case statements are great for switching on type where necessary. Because `StandardError` is a superclass of most other errors, the reordering changes the case meaning.

So for me, where the benefits are so context-sensitive, it seems like it's not a risk worth running. It might be a performance benefit, it might not; a user will need to profile to know either way, so what's the cop for? If a condition reordered to the top of a case statement to comply with this cop is expensive, the performance effects could be significantly negative; rubocop simply has no way of knowing.

> Like many of the other performance related cops, the performance benefits are highly dependent on the context of the code itself.

To be completely honest, for exactly this reason I'm pretty skeptical that a linter should be providing performance advice at all, as it can't know whether its advice will help or not. In my experience, performance optimisations gleaned from one codebase very rarely confer significant advantages in others; you always, always need to profile.

While I'd much prefer removal, I wouldn't object too strenuously to disabling it by default. I'm not sure I see the value in this option; a user would presumably need to know the cop was beneficial to want to enable it, implying they'd profiled the code to find out. I don't think updating the docs is sufficient on its own, but I do think we should do this if the cop is disabled.

**Comment by rrosenblum** [comment]
@urbanautomaton you bring up some great points. 

In general, I feel as though a cop should be removed if it doesn't work, doesn't provide value, or is no longer functional. In this case, there is an unsafe auto-correct, but the cop itself still works and can provide value to someone. 

Unfortunately, the average user likely only sees the output message from the offense. They likely don't read the documentation for the actual cops. In this case, disabling auto-correct by default would force the user to make decisions about their code themselves. I think the best bet would be to disable auto-correct or the cop by default. Either way, we should look at updating the documentation for the cop to make sure all risks are known.

> To be completely honest, for exactly this reason I'm pretty skeptical that a linter should be providing performance advice at all, as it can't know whether its advice will help or not. In my experience, performance optimisations gleaned from one codebase very rarely confer significant advantages in others; you always, always need to profile.

All of the performance cops have been a touchy subject for this reason. There are plans to move the performance cops out into a separate gem which would effectively disable them all by default. It is unlikely that the performance cops will make a slow application fast, however it allows us to recommend, and auto-correct, improvements for the user. 

`RedundantMerge` is a good example of this. `hash.merge!(a: 1)` vs `hash[:a] = 1` by itself is a negligible change. The performance gain for a single change is small. The readability of both pieces of code is essentially the same. I find it beneficial to be able to provide improvements like this.

**Comment by urbanautomaton** [comment]
> In this case, there is an unsafe auto-correct, but the cop itself still works...

I don't agree that the auto-correct is the problem - the problem is that the edit required to pass the cop is unsafe whether automatically applied or not, and can never be made safe. So I'd argue that the cop doesn't (and can't) work.

> Unfortunately, the average user likely only sees the output message from the offense. They likely don't read the documentation for the actual cops. In this case, disabling auto-correct by default would force the user to make decisions about their code themselves. I think the best bet would be to disable auto-correct or the cop by default.

If we do retain the cop, I think it should be disabled by default, rather than just turning off autocorrect. Rubocop speaks with a normative voice (its examples are all "good" vs "bad"), so an inexperienced user will likely take an error as an instruction and implement it, autocorrect or no. A linter where you constantly have to evaluate whether the advice is safe loses much of its value.

> All of the performance cops have been a touchy subject for this reason. There are plans to move the performance cops out into a separate gem which would effectively disable them all by default.

Ah, that's really interesting, thanks for the info. I'll stop there though cos I think I've dragged us off-topic - maybe one for the mailing list. 😄

**Comment by bbatsov** [comment]
Closing in favour of #6188.
